### ISSUE TITLE
Add assistance to detect common backwards compatibilty issues

### ISSUE THREAD

**Issue by jsoriano** [elastic/elastic-package#579]
It could be nice to have some feature to assist detecting backwards compatibility issues in packages. This could be part of `elastic-package check`, but it would have the particularity of needing to download (or checkout) previous versions of the package, so maybe it should be optional.

There are some things that could be automatically checked between versions:
* Incompatible changes in field data types.
* Removed fields.
* Changes in sets of dimensions.
* Changes in policy changes, or removed variables.

**Comment by mtojek** [comment]
The idea is nice of course, but I don't see an easy way to evaluate these changes. A package is considered to be incompatible if we can't install it or tests are failing, so we could just remodel the testing pipeline and run tests for more stacks - this is the dumb way. To implement the smartest one, we need to collect some ideas.

**Comment by jsoriano** [comment]
> The idea is nice of course, but I don't see an easy way to evaluate these changes. A package is considered to be incompatible if we can't install it or tests are failing, so we could just remodel the testing pipeline and run tests for more stacks

These checks wouldn't be related to multiple versions of the stack, but to multiple versions of the package itself. The E2E way to do it would be to install an old version of the package, upgrade to the new one, and do something as running the system tests, and check that everything works. But I think that many things can be "statically" detected.

For example:
* Comparing old and new list of fields we can detect that a field has been removed, or changed to an incompatible data type. We can also detect if the set of dimensions has changed. 
* Comparing old and new variables we can check if a new one has been added without a default, what could complicate automatic upgrades. We can also check if variables have been removed, or if they are not used.

I think that this kind of issues are actually detected more reliably with static checks, other kind of tests as E2E or system tests may not use all fields, or variables, and then they wouldn't detect issues with them. For example incompatible changes on types can be easy to detect by comparing old and new fields definitions, these issues may not cause any issue during ingestion, but can cause problems in Kibana.

Other breaking changes as the ones in policy templates may be more complicated, or not possible to detect without system tests, but we can incrementally add checks later if we find ways to implement them.

**Comment by droberts195** [comment]
> But I think that many things can be "statically" detected.

The transform upgrade process that @qn895 is working on would also benefit from static checks against the previous package version. Since package upgrades cannot delete indices we're making a rule that if you change a transform in any way then you have to give it a different destination index, as it's too hard to be sure if the new transform and old destination index would be compatible. So ideally we'd validate that this rule has been followed rather than risking problems being found a while after install of the new package.

**Comment by jlind23** [comment]
@jsoriano @mrodm Shall we also piggy back on this for packages upgrade? We have seen failures where a package type was changed to an input package type and it broke the package upgrade. Relying on CI to test packages installation AND upgrade from the previous patch/minor would definitely be helpful.
cc @kpollich

**Comment by jsoriano** [comment]
> We have seen failures where a package type was changed to an input package type and it broke the package upgrade.

Changing the package type from integration to input should not be breaking, but yes, cases like this one should be detected at some testing stage.

**Comment by lucabelluccini** [comment]
Hello, I have some questions:
- Do we have plans to ensure 3rd party services logs are tested weekly or nightly and the indexing still works? Example: reading data from a SaaS service and they change the format of the logs.
- Do we have nightlies to ensure fields are kept of the same type across upgrades? Note we might need to do this both for package upgrades and also in minor Fleet upgrades, as the structure of the Index Template can be changed by Fleet.

**Comment by jsoriano** [comment]
> * Do we have plans to ensure 3rd party services logs are tested weekly or nightly and the indexing still works? Example: reading data from a SaaS service and they change the format of the logs.

I don't think so. Actually integrations for some propietary sources are usually tested with mocked data. Package developers need to explicitly check if formats have changed. There is an open issue focused on finding changes in [non-SaaS] services: https://github.com/elastic/elastic-package/issues/1902.

In any case I would consider breaking changes in monitored services a separate issue. Here we intend to prevent the introduction of breaking changes in the package itself.

> * Do we have nightlies to ensure fields are kept of the same type across upgrades? Note we might need to do this both for package upgrades and also in minor Fleet upgrades, as the structure of the Index Template can be changed by Fleet.

Not for changes in packages, this is the kind of check that we would introduce by solving this task.

For changes in the stack that might produce issues with packages, we have https://buildkite.com/elastic/integrations-schedule-daily/, where we test against latest versions of 7.x, 8.x and serverless, though we don't test upgrades.

Some changes in Fleet might be tested in Kibana CI itself, we have probably already some tests around there.
